File: test/performance/graphs/xsymm.r

# ==================================================================================================
# This file is part of the CLBlast project. The project is licensed under Apache Version 2.0. This
# project uses a tab-size of two spaces and a max-width of 100 characters per line.
#
# Author(s):
#   Cedric Nugteren <www.cedricnugteren.nl>
#
# This file implements the performance script for the Xsymm routine
#
# ==================================================================================================

# Includes the common functions
args <- commandArgs(trailingOnly = FALSE)
thisfile <- (normalizePath(sub("--file=", "", args[grep("--file=", args)])))
source(file.path(dirname(thisfile), "common.r"))

# ==================================================================================================

# Settings
routine_name <- "xsymm"
parameters <- c("-m","-n","-layout","-triangle","-side",
                "-num_steps","-step","-runs","-precision")
precision <- 32

# Sets the names of the test-cases
test_names <- list(
  "multiples of 128",
  "multiples of 128 (+1)",
  "around m=n=512",
  "around m=n=2048",
  "layouts and triangle/side (m=n=1024)",
  "powers of 2"
)

# Defines the test-cases
test_values <- list(
  list(c(128, 128, 0, 0, 0, 16, 128, num_runs, precision)),
  list(c(129, 129, 0, 0, 0, 16, 128, num_runs, precision)),
  list(c(512, 512, 0, 0, 0, 16, 1, num_runs, precision)),
  list(c(2048, 2048, 0, 0, 0, 16, 1, num_runs, precision)),
  list(
    c(1024, 1024, 0, 0, 0, 1, 0, num_runs, precision),
    c(1024, 1024, 0, 0, 1, 1, 0, num_runs, precision),
    c(1024, 1024, 0, 1, 0, 1, 0, num_runs, precision),
    c(1024, 1024, 0, 1, 1, 1, 0, num_runs, precision),
    c(1024, 1024, 1, 0, 0, 1, 0, num_runs, precision),
    c(1024, 1024, 1, 0, 1, 1, 0, num_runs, precision),
    c(1024, 1024, 1, 1, 0, 1, 0, num_runs, precision),
    c(1024, 1024, 1, 1, 1, 1, 0, num_runs, precision)
  ),
  list(
    c(8, 8, 0, 0, 0, 1, 0, num_runs, precision),
    c(16, 16, 0, 0, 0, 1, 0, num_runs, precision),
    c(32, 32, 0, 0, 0, 1, 0, num_runs, precision),
    c(64, 64, 0, 0, 0, 1, 0, num_runs, precision),
    c(128, 128, 0, 0, 0, 1, 0, num_runs, precision),
    c(256, 256, 0, 0, 0, 1, 0, num_runs, precision),
    c(512, 512, 0, 0, 0, 1, 0, num_runs, precision),
    c(1024, 1024, 0, 0, 0, 1, 0, num_runs, precision),
    c(2048, 2048, 0, 0, 0, 1, 0, num_runs, precision),
    c(4096, 4096, 0, 0, 0, 1, 0, num_runs, precision),
    c(8192, 8192, 0, 0, 0, 1, 0, num_runs, precision)
  )
)

# Defines the x-labels corresponding to the test-cases
test_xlabels <- list(
  "matrix sizes (m=n)",
  "matrix sizes (m=n)",
  "matrix sizes (m=n)",
  "matrix sizes (m=n)",
  "layout (row/col), triangle (up/lo), side (l/r)",
  "matrix sizes (m=n)"
)

# Defines the x-axis of the test-cases
test_xaxis <- list(
  c("m", ""),
  c("m", ""),
  c("m", ""),
  c("m", ""),
  list(1:8, c("row,up,l", "row,up,r", "row,lo,l", "row,lo,r",
              "col,up,l", "col,up,r", "col,lo,l", "col,lo,r")),
  c("m", "x")
)

# ==================================================================================================

# Start the script
main(routine_name=routine_name, precision=precision, test_names=test_names, test_values=test_values,
     test_xlabels=test_xlabels, test_xaxis=test_xaxis, metric_gflops=TRUE)

# ==================================================================================================

File: src/routines/xsymm.cc

// =================================================================================================
// This file is part of the CLBlast project. The project is licensed under Apache Version 2.0. This
// project loosely follows the Google C++ styleguide and uses a tab-size of two spaces and a max-
// width of 100 characters per line.
//
// Author(s):
//   Cedric Nugteren <www.cedricnugteren.nl>
//
// This file implements the Xsymm class (see the header for information about the class).
//
// =================================================================================================

#include "internal/routines/xsymm.h"

#include <string>
#include <vector>

namespace clblast {
// =================================================================================================

// Constructor: forwards to base class constructor
template <typename T>
Xsymm<T>::Xsymm(CommandQueue &queue, Event &event):
    Xgemm<T>(queue, event) {
}

// =================================================================================================

// The main routine
template <typename T>
StatusCode Xsymm<T>::DoSymm(const Layout layout, const Side side, const Triangle triangle,
                            const size_t m, const size_t n,
                            const T alpha,
                            const Buffer &a_buffer, const size_t a_offset, const size_t a_ld,
                            const Buffer &b_buffer, const size_t b_offset, const size_t b_ld,
                            const T beta,
                            const Buffer &c_buffer, const size_t c_offset, const size_t c_ld) {

  // Makes sure all dimensions are larger than zero
  if ((m == 0) || (n == 0) ) { return StatusCode::kInvalidDimension; }

  // Computes the k dimension. This is based on whether or not the symmetric matrix is A (on the
  // left) or B (on the right) in the Xgemm routine.
  size_t k = (side == Side::kLeft) ? m : n;

  // Checks for validity of the squared A matrix
  auto status = TestMatrixA(k, k, a_buffer, a_offset, a_ld, sizeof(T));
  if (ErrorIn(status)) { return status; }

  // Determines which kernel to run based on the layout (the Xgemm kernel assumes column-major as
  // default) and on whether we are dealing with an upper or lower triangle of the symmetrix matrix
  bool is_upper = ((triangle == Triangle::kUpper && layout != Layout::kRowMajor) ||
                   (triangle == Triangle::kLower && layout == Layout::kRowMajor));
  auto kernel_name = (is_upper) ? "SymmUpperToSquared" : "SymmLowerToSquared";

  // Temporary buffer for a copy of the symmetric matrix
  try {
    auto temp_symm = Buffer(context_, CL_MEM_READ_WRITE, k*k*sizeof(T));

    // Creates a general matrix from the symmetric matrix to be able to run the regular Xgemm
    // routine afterwards
    try {
      auto& program = GetProgramFromCache();
      auto kernel = Kernel(program, kernel_name);

      // Sets the arguments for the symmetric-to-squared kernel
      kernel.SetArgument(0, static_cast<int>(k));
      kernel.SetArgument(1, static_cast<int>(a_ld));
      kernel.SetArgument(2, static_cast<int>(a_offset));
      kernel.SetArgument(3, a_buffer());
      kernel.SetArgument(4, static_cast<int>(k));
      kernel.SetArgument(5, static_cast<int>(k));
      kernel.SetArgument(6, static_cast<int>(0));
      kernel.SetArgument(7, temp_symm());

      // Uses the common padding kernel's thread configuration. This is allowed, since the
      // symmetry-to-squared kernel uses the same parameters.
      auto global = std::vector<size_t>{Ceil(CeilDiv(k, db_["PAD_WPTX"]), db_["PAD_DIMX"]),
                                        Ceil(CeilDiv(k, db_["PAD_WPTY"]), db_["PAD_DIMY"])};
      auto local = std::vector<size_t>{db_["PAD_DIMX"], db_["PAD_DIMY"]};
      status = RunKernel(kernel, global, local);
      if (ErrorIn(status)) { return status; }

      // Runs the regular Xgemm code with either "C := AB+C" or ...
      if (side == Side::kLeft) {
        status = DoGemm(layout, Transpose::kNo, Transpose::kNo,
                        m, n, k,
                        alpha,
                        temp_symm, 0, k,
                        b_buffer, b_offset, b_ld,
                        beta,
                        c_buffer, c_offset, c_ld);
      }

      // ... with "C := BA+C". Note that A and B are now reversed.
      else {
        status = DoGemm(layout, Transpose::kNo, Transpose::kNo,
                        m, n, k,
                        alpha,
                        b_buffer, b_offset, b_ld,
                        temp_symm, 0, k,
                        beta,
                        c_buffer, c_offset, c_ld);

        // A and B are now reversed, so also reverse the error codes returned from the Xgemm routine
        switch(status) {
          case StatusCode::kInvalidMatrixA:      status = StatusCode::kInvalidMatrixB; break;
          case StatusCode::kInvalidMatrixB:      status = StatusCode::kInvalidMatrixA; break;
          case StatusCode::kInvalidLeadDimA:     status = StatusCode::kInvalidLeadDimB; break;
          case StatusCode::kInvalidLeadDimB:     status = StatusCode::kInvalidLeadDimA; break;
          case StatusCode::kInsufficientMemoryA: status = StatusCode::kInsufficientMemoryB; break;
          case StatusCode::kInsufficientMemoryB: status = StatusCode::kInsufficientMemoryA; break;
        }
      }

      // Return the status of the Xgemm routine
      return status;
    } catch (...) { return StatusCode::kInvalidKernel; }
  } catch (...) { return StatusCode::kTempBufferAllocFailure; }
}

// =================================================================================================

// Compiles the templated class
template class Xsymm<float>;
template class Xsymm<double>;
template class Xsymm<float2>;
template class Xsymm<double2>;

// =================================================================================================
} // namespace clblast
